packages/s3-cache/src/index.ts

```diff
@@ -38,21 +44,58 @@ export class S3Cache implements CacheStrategy {
   }
 
   async set(pageKey: string, cacheKey: string, data: CacheEntry): Promise<void> {
-    const input = {
+    const baseInput = {
       Bucket: this.bucketName,
-      Key: `${pageKey}/${cacheKey}`,
-      ...(data.tags?.length ? { Metadata: { tags: data.tags.join(TAGS_SEPARATOR) } } : {})
+      Key: `${pageKey}/${cacheKey}`
     }
 
     if (data.value?.kind === 'PAGE') {
-      await this.client.putObject({ ...input, Key: `${input.Key}.html`, Body: data.value.html })
+      const headersTags = this.buildTagKeys(data.value.headers?.[NEXT_CACHE_TAGS_HEADER]?.toString())
+      const input = { ...baseInput, ...(headersTags ? { Tagging: headersTags } : {}) }
+      await this.client.putObject({
+        ...input,
+        Key: `${input.Key}.html`,
+        Body: data.value.html
+      })
+      await this.client.putObject({
+        ...input,
+        Key: `${input.Key}.json`,
+        Body: JSON.stringify(data)
+      })
+      return
     }
 
-    await this.client.putObject({ ...input, Key: `${input.Key}.json`, Body: JSON.stringify(data) })
+    await this.client.putObject({
+      ...baseInput,
+      Key: `${baseInput.Key}.json`,
+      Body: JSON.stringify(data),
+      ...(data.tags?.length ? { Tagging: `${this.buildTagKeys(data.tags)}` } : {})
+    })
   }
 
-  async revalidateTag(): Promise<void> {
-    // TODO add revalidate tag functionality for s3
+  async revalidateTag(tag: string): Promise<void> {
+    let nextContinuationToken: string | undefined = undefined
+    do {
+      const { Contents: contents = [], NextContinuationToken: token }: ListObjectsV2CommandOutput =
+        await this.client.listObjectsV2({
+          Bucket: this.bucketName,
+          ContinuationToken: nextContinuationToken
+        })
+      nextContinuationToken = token
+
+      for (const { Key: key } of contents) {
+        if (!key) continue
+
+        const args = { Bucket: this.bucketName, Key: key }
+        const { TagSet = [] } = await this.client.getObjectTagging(args)
+
+        const tags = TagSet.filter(({ Key: key }) => key?.startsWith(TAG_PREFIX)).map(({ Value: tags }) => tags || '')
+
+        if (tags.includes(tag)) {
+          await this.client.deleteObject(args)
```

Comment:
instead if single deleteObject, let's use `deleteObjects`, so we can delete multiple files via single request, we can send up 1k keys to delete, so we can minimize number of requests what we do to s3

Reply:
😓 Thanks, I reworked it based on `deleteObjects`